src/mlpack/methods/reinforcement_learning/async_learning_impl.hpp

```diff
@@ -0,0 +1,164 @@
+{ /* Nothing to do here. */ };
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+template <typename Measure>
+void AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::Train(Measure& measure)
+{
+  /**
+   * OpenMP doesn't support shared class member variables.
+   * So we need to copy them to local variables.
+   */
+  NetworkType learningNetwork = std::move(this->learningNetwork);
+  if (learningNetwork.Parameters().is_empty())
+    learningNetwork.ResetParameters();
+  NetworkType targetNetwork = learningNetwork;
+  size_t totalSteps = 0;
+  PolicyType policy = this->policy;
+  bool stop = false;
+
+  // Set up worker pool, worker 0 will be deterministic for evaluation.
+  std::vector<WorkerType> workers;
+  for (size_t i = 0; i <= config.NumWorkers(); ++i)
+  {
+    workers.push_back(WorkerType(updater, environment, config, !i));
+    workers.back().Initialize(learningNetwork);
+  }
+  // Set up task queue corresponding to worker pool.
+  std::queue<size_t> tasks;
+  for (size_t i = 0; i <= config.NumWorkers(); ++i)
+    tasks.push(i);
+
+  // Synchronization for get/put task.
+  #ifdef HAS_OPENMP
+    omp_lock_t tasksLock;
+    omp_init_lock(&tasksLock);
+  #endif
+
+  /**
+   * Compute the number of threads for the for-loop. In general, we should use
+   * OpenMP task rather than for-loop, here we do so to be compatible with some
+   * compiler. We can switch to OpenMP task once MSVC supports OpenMP 3.0.
+   */
+  size_t numThreads = 0;
+  #ifdef HAS_OPENMP
+    #pragma omp parallel reduction(+:numThreads)
+  #endif
+  numThreads++;
+  Log::Debug << numThreads << " threads will be used in total." << std::endl;
+
+  #ifdef HAS_OPENMP
+    #pragma omp parallel for shared(stop, workers, tasks, learningNetwork, \
+        targetNetwork, totalSteps, policy)
+  #endif
+  for (omp_size_t i = 0; i < numThreads; ++i)
+  {
+    #ifdef HAS_OPENMP
+      #pragma omp critical
+    #endif
+    {
+      Log::Debug << "Thread " << omp_get_thread_num() <<
+          " started." << std::endl;
+    }
+    while (!stop)
+    {
+      // Assign task to current thread from queue.
+      #ifdef HAS_OPENMP
+        omp_set_lock(&tasksLock);
+        if (tasks.empty())
+        {
+          omp_unset_lock(&tasksLock);
+          continue;
+        }
+      #endif
+      size_t task = tasks.front();
+      tasks.pop();
+      #ifdef HAS_OPENMP
+        omp_unset_lock(&tasksLock);
+      #endif
+
+      // Get corresponding worker.
+      WorkerType& worker = workers[task];
+      double episodeReturn;
+      if (worker.Step(learningNetwork, targetNetwork, totalSteps,
+          policy, episodeReturn) && !task)
+      {
+        stop = measure(episodeReturn);
+      }
+
+      // Put task back to queue.
+      #ifdef HAS_OPENMP
+        omp_set_lock(&tasksLock);
+      #endif
+      tasks.push(task);
```

Comment:
Ah, I see what you mean.  You could possibly restructure something like...

```
bool hasTask = false;

#pragma omp parallel ...
while (!stop) // worker loop
{
  // Push old task if needed, get new task.
  #pragma omp critical
  {
    if (hasTask)
      tasks.push(task);
    task = tasks.front();
    tasks.pop();
    hasTask = true;
  }

  // Do task...
}
```

If this is possible, I think it removes all need for `omp_set_lock()` and `omp_unset_lock()`.  What do you think?  It's possible I overlooked something.